crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+
+        let mut bc = GuestBorrows::new();
+        let as_str = path.as_raw(&mut bc)?;
+        let path = unsafe { &*as_str };
+
+        trace!("     | (path_ptr,path_len)='{}'", path);
+
+        let rights = types::Rights::PATH_OPEN | types::Rights::PATH_CREATE_DIRECTORY;
+        let entry = unsafe { self.get_entry(dirfd)? };
+        let resolved = path::get(
+            &entry,
+            rights,
+            types::Rights::empty(),
+            types::Lookupflags::empty(),
+            path,
+            false,
+        )?;
+        resolved.create_directory()
+    }
+
+    fn path_filestat_get(
+        &self,
+        dirfd: types::Fd,
+        flags: types::Lookupflags,
+        path: &GuestPtr<'_, str>,
+    ) -> Result<types::Filestat> {
+        trace!(
+            "path_filestat_get(dirfd={:?}, flags={:?}, path={:?})",
+            dirfd,
+            flags,
+            path,
+        );
+
+        let mut bc = GuestBorrows::new();
+        let as_str = path.as_raw(&mut bc)?;
+        let path = unsafe { &*as_str };
+
+        trace!("     | (path_ptr,path_len)='{}'", path);
+
+        let entry = unsafe { self.get_entry(dirfd)? };
+        let resolved = path::get(
+            &entry,
+            types::Rights::PATH_FILESTAT_GET,
+            types::Rights::empty(),
+            flags,
+            path,
+            false,
+        )?;
+        let host_filestat = match resolved.dirfd() {
+            Descriptor::VirtualFile(virt) => virt
+                .openat(
+                    std::path::Path::new(resolved.path()),
+                    false,
+                    false,
+                    types::Oflags::empty(),
+                    types::Fdflags::empty(),
+                )?
+                .filestat_get()?,
+            _ => path::filestat_get(resolved, flags)?,
+        };
+
+        trace!("     | *filestat_ptr={:?}", host_filestat);
+
+        Ok(host_filestat)
+    }
+
+    fn path_filestat_set_times(
+        &self,
+        dirfd: types::Fd,
+        flags: types::Lookupflags,
+        path: &GuestPtr<'_, str>,
+        atim: types::Timestamp,
+        mtim: types::Timestamp,
+        fst_flags: types::Fstflags,
+    ) -> Result<()> {
+        trace!(
+            "path_filestat_set_times(dirfd={:?}, flags={:?}, path={:?}, atim={}, mtim={}, fst_flags={})",
+            dirfd,
+            flags,
+            path,
+            atim,
+            mtim,
+            fst_flags,
+        );
+
+        let mut bc = GuestBorrows::new();
+        let as_str = path.as_raw(&mut bc)?;
+        let path = unsafe { &*as_str };
+
+        trace!("     | (path_ptr,path_len)='{}'", path);
+
+        let entry = unsafe { self.get_entry(dirfd)? };
+        let resolved = path::get(
+            &entry,
+            types::Rights::PATH_FILESTAT_SET_TIMES,
+            types::Rights::empty(),
+            flags,
+            path,
+            false,
+        )?;
```

Comment:
Could the idiom of unsafely getting a `&str` and then calling `path::get` be wrapped up in a safe function to `unsafe` doesn't have to be used each time a path is resolved?

Reply:
I've done just that. Now `path::get` takes `&GuestPtr<'_, str>` directly and does borrowing internally. This actually, at the surface at least, gets rid of the double-immutable-borrow problem.